File: 여름청 강수량/arima.py
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.arima.model import ARIMA

# CSV 파일에서 데이터를 읽습니다.
filename = 'smoothed_data.csv'  # CSV 파일의 경로
data = pd.read_csv(filename, encoding='utf-8')  # UTF-8 인코딩을 사용

# 연도와 스무딩된 강수량 데이터를 추출합니다.
years = data['Year']
smoothed_rainfall = data['SmoothedRainfall']

# 스무딩된 강수량 데이터를 시각화합니다.
plt.figure()
plt.plot(years, smoothed_rainfall, label='Smoothed Rainfall Data')
plt.title('Smoothed Rainfall Data')
plt.xlabel('Year')
plt.ylabel('Rainfall')
plt.legend()
plt.grid(True)
plt.show()

# ARIMA(1,2,1) 모델 설정 및 추정
model = ARIMA(smoothed_rainfall, order=(3, 2, 5))
model_fit = model.fit()

# 모델 요약 정보 출력
print(model_fit.summary())

# 예측 기간 설정
horizon = 10  # 10년 예측

# 예측 수행
forecast = model_fit.get_forecast(steps=horizon)
forecast_mean = forecast.predicted_mean
conf_int = forecast.conf_int()

# 예측 결과를 시각화하기 위한 미래 연도 계산
future_years = years.iloc[-1] + pd.Series(range(1, horizon + 1))

# 예측 결과 시각화
plt.figure()
plt.plot(years, smoothed_rainfall, label='Smoothed Rainfall Data')
plt.plot(future_years, forecast_mean, 'r--', label='ARIMA(3,2,5) Forecast')
plt.fill_between(future_years, conf_int['lower SmoothedRainfall'], conf_int['upper SmoothedRainfall'], color='gray', alpha=0.2, label='95% Confidence Interval')
plt.title('ARIMA(3,2,5) Model Forecast')
plt.xlabel('Year')
plt.ylabel('Rainfall')
plt.legend()
plt.grid(True)
plt.show()

File: 여름청 강수량/arima.m
% CSV 파일을 읽어서 데이터셋을 가져옵니다.
filename = '장마철 강수량 연도별 데이터.csv';  % CSV 파일의 경로를 지정하세요.
data = readtable(filename);  % 데이터셋을 읽어옵니다.

% 데이터에서 연도와 강수량 데이터를 추출합니다.
years = data{:, 1};  % 첫 번째 열은 연도
rainfall = data{:, 2};  % 두 번째 열은 강수량

% 이동 평균 스무딩을 적용합니다.
% 윈도우 크기를 조정해서 적절한 스무딩 효과를 얻을 수 있습니다.
windowSize = 5;  % 이동 평균의 윈도우 크기 (예: 5)
smoothed_rainfall = movmean(rainfall, windowSize);

% ARIMA(1,2,1) 모델을 설정합니다.
model = arima(1, 2, 1);

% ARIMA 모델을 스무딩된 데이터에 적합시킵니다.
fit = estimate(model, smoothed_rainfall);

% 예측을 수행합니다.
num_forecast_periods = 10;  % 예측할 기간(단위: 연도)을 설정합니다.
[forecast, forecast_mse] = forecast(fit, num_forecast_periods);

% 예측 시각화
figure;
% 원본 데이터 시각화
plot(years, smoothed_rainfall, 'b', 'DisplayName', '스무딩된 데이터');
hold on;
% 예측 데이터 시각화
plot(years(end) + (1:num_forecast_periods), forecast, 'r', 'DisplayName', '예측 데이터');

% 그래프 설정
xlabel('연도');
ylabel('강수량');
title('ARIMA(1,2,1) 모델을 사용한 스무딩된 강수량 예측');
legend;
grid on;

% 예측 신뢰 구간을 계산합니다.
forecast_sd = sqrt(forecast_mse);
forecast_upper = forecast + 1.96 * forecast_sd;
forecast_lower = forecast - 1.96 * forecast_sd;

% 신뢰 구간을 시각화
plot(years(end) + (1:num_forecast_periods), forecast_upper, 'r--', 'DisplayName', '95% 신뢰 구간 상한');
plot(years(end) + (1:num_forecast_periods), forecast_lower, 'r--', 'DisplayName', '95% 신뢰 구간 하한');

hold off;
